src/optimizer/join_order/cardinality_estimator.cpp

```diff
@@ -432,11 +432,13 @@ void CardinalityEstimator::UpdateTotalDomains(JoinNode &node, LogicalOperator &o
 			}
 		}
 
+		bool have_stats = false;
 		if (catalog_table && actual_binding != relation_column_to_original_column.end()) {
```

Comment:
Thank you Tom for picking this up so quickly. One little concern I have is that when the catalog table is not a `DuckTable` then `catalog_table` is nullptr because it is being extracted from `LogicalGet`'a GetTable, which then goes and looks for a hardcode DuckTable (and friends) dynamic cast. This make this code unfriendly to other catalog implementations. IMO (but I'm deferring to the experts), `GetTable` on `LogicalGet` should go away or have another implementation. Here, maybe, you can ask the table function if it has statistics and use that instead of  using `catalog_table` as a boolean.

Reply:
`GetLogicalGet()` is meant to traverse the plan and find the DataSource operator where the data is being pulled from. When the operator is found, the previous implementation was to call `GetCatalogTable`. Just as you mentioned, this is unfriendly to other catalog implementations, so line 438 calls the table statistics function. 

The reason the catalog table check is there is to avoid join order regressions, however, I'm not opposed to taking it out to avoid the issue of duck tables vs. other catalog implementations. It's also possible that some of the variable names are a bit confusing, as the `get` variable can be any one of the [Data Source](https://github.com/duckdb/duckdb/blob/08cb5cc819565e1e94ecac315a31237aea970868/src/include/duckdb/common/enums/logical_operator_type.hpp#L37) operators or `Logical_projection` (in a few select cases).

If @Mytherin thinks it's a good idea to remove the check for the catalog_table and accept the join order regressions, then I'm happy to remove the check all together.

Reply:
Ah, but I think I understand more now, any other catalog implementation won't get any statistics ever during the join ordering process, and `catalog_table` will always be null. To be friendly to other catalog implementations the check should be removed all together yes